// @ngrx/effects/migrations/9_0_0/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@ngrx/effects/migrations/9_0_0/index", ["require", "exports", "@angular-devkit/schematics", "@ngrx/effects/schematics-core", "@ngrx/effects/schematics-core", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const schematics_1 = require("@angular-devkit/schematics");
    const schematics_core_1 = require("@ngrx/effects/schematics-core");
    const schematics_core_2 = require("@ngrx/effects/schematics-core");
    const ts = require("typescript");
    function renameErrorHandlerConfig() {
        return (tree, ctx) => {
            schematics_core_1.visitTSSourceFiles(tree, sourceFile => {
                const changes = replaceEffectConfigKeys(sourceFile, 'resubscribeOnError', 'useEffectsErrorHandler');
                schematics_core_1.commitChanges(tree, sourceFile.fileName, changes);
                if (changes.length) {
                    ctx.logger.info(`[@ngrx/effects] Updated Effects configuration, see the migration guide (https://ngrx.io/guide/migration/v9#effects) for more info`);
                }
            });
        };
    }
    function replaceEffectConfigKeys(sourceFile, oldText, newText) {
        const changes = [];
        ts.forEachChild(sourceFile, node => {
            visitCreateEffectFunctionCreator(node, createEffectNode => {
                const [effectDeclaration, configNode] = createEffectNode.arguments;
                if (configNode) {
                    findAndReplaceText(configNode);
                }
            });
            visitEffectDecorator(node, effectDecoratorNode => {
                findAndReplaceText(effectDecoratorNode);
            });
        });
        return changes;
        function findAndReplaceText(node) {
            visitIdentifierWithText(node, oldText, match => {
                changes.push(schematics_core_2.createReplaceChange(sourceFile, match, oldText, newText));
            });
        }
    }
    function visitIdentifierWithText(node, text, visitor) {
        if (ts.isIdentifier(node) && node.text === text) {
            visitor(node);
        }
        ts.forEachChild(node, childNode => visitIdentifierWithText(childNode, text, visitor));
    }
    function visitEffectDecorator(node, visitor) {
        if (ts.isDecorator(node) &&
            ts.isCallExpression(node.expression) &&
            ts.isIdentifier(node.expression.expression) &&
            node.expression.expression.text === 'Effect') {
            visitor(node);
        }
        ts.forEachChild(node, childNode => visitEffectDecorator(childNode, visitor));
    }
    function visitCreateEffectFunctionCreator(node, visitor) {
        if (ts.isCallExpression(node) &&
            ts.isIdentifier(node.expression) &&
            node.expression.text === 'createEffect') {
            visitor(node);
        }
        ts.forEachChild(node, childNode => visitCreateEffectFunctionCreator(childNode, visitor));
    }
    function default_1() {
        return schematics_1.chain([renameErrorHandlerConfig()]);
    }
    exports.default = default_1;
});
