// jasmine-spec-reporter/built/processors/pretty-errors-processor.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var configuration_1 = require("../configuration");
var display_processor_1 = require("../display-processor");
var STACK_REG_EXP = /\((.*):(\d+):(\d+)\)/;
var CONTEXT = 2;
var PrettyErrorsProcessor = /** @class */ (function (_super) {
    __extends(PrettyErrorsProcessor, _super);
    function PrettyErrorsProcessor() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PrettyErrorsProcessor.prototype.displaySpecErrorMessages = function (spec, log) {
        return this.configuration.spec.displayStacktrace === configuration_1.StacktraceOption.PRETTY ? this.displayErrorMessages(spec) : log;
    };
    PrettyErrorsProcessor.prototype.displaySummaryErrorMessages = function (spec, log) {
        return this.configuration.summary.displayStacktrace === configuration_1.StacktraceOption.PRETTY ? this.displayErrorMessages(spec) : log;
    };
    PrettyErrorsProcessor.prototype.displayErrorMessages = function (spec) {
        var logs = [];
        for (var _i = 0, _a = spec.failedExpectations; _i < _a.length; _i++) {
            var failedExpectation = _a[_i];
            logs.push("- ".failed + failedExpectation.message.failed);
            if (failedExpectation.stack) {
                logs.push(this.prettifyStack(failedExpectation.stack));
            }
        }
        return logs.join("\n");
    };
    PrettyErrorsProcessor.prototype.prettifyStack = function (stack) {
        var _this = this;
        var logs = [];
        var filteredStack = this.configuration.stacktrace.filter(stack);
        var stackRegExp = new RegExp(STACK_REG_EXP);
        filteredStack.split("\n").forEach(function (stackLine) {
            if (stackRegExp.test(stackLine)) {
                var _a = stackLine.match(stackRegExp), filename = _a[1], lineNb = _a[2], columnNb = _a[3];
                var errorContext = _this.retrieveErrorContext(filename, parseInt(lineNb, 10), parseInt(columnNb, 10));
                logs.push(filename.cyan + ":" + lineNb.yellow + ":" + columnNb.yellow);
                logs.push(errorContext);
            }
        });
        return "\n" + logs.join("\n") + "\n";
    };
    PrettyErrorsProcessor.prototype.retrieveErrorContext = function (filename, lineNb, columnNb) {
        var logs = [];
        var fileLines = fs.readFileSync(filename, "utf-8")
            .split("\n");
        for (var i = 0; i < fileLines.length; i++) {
            var errorLine = lineNb - 1;
            if (i >= errorLine - CONTEXT && i <= errorLine + CONTEXT) {
                logs.push(fileLines[i]);
            }
            if (i === errorLine) {
                logs.push(" ".repeat(columnNb - 1) + "~".red);
            }
        }
        return logs.join("\n");
    };
    return PrettyErrorsProcessor;
}(display_processor_1.DisplayProcessor));
exports.PrettyErrorsProcessor = PrettyErrorsProcessor;
